File: DengJH/python/ConnectedComponent.py
"""
连通域分析: 查找连通域, 并染色, 处理对象通常是一张二值化的图像(代码使用0和255), 按行或按列扫描图像所有像素, 前景像素为1, 背景像素为0,
label从2开始计数, 像素通常有4邻域或8邻域.

two-pass法: 扫描2遍图像, 第一遍扫描时赋予每个像素位置一个label, 扫描过程中同一个连通域内的像素集合中可能被
赋予一个或多个不同的label, 因此需要将这些属于同一个连通域但具有不同值的label合并, 这个功能由第二遍扫描(第二遍为逆序扫描)完成, 即
为一个连通域赋予一个相同的label(通常是该连通域中最小的label)

seed filling(种子填充法): 选取一个前景像素点作为种子, 然后根据连通域两个基本条件(像素相同,位置相邻), 将与种子相邻的前景像素合并到同一个像素集合中, 最后得到该像素集合则为一个连通区域
"""

import cv2
import numpy as np

# 4邻域的连通域和 8邻域的连通域
# [row, col]
NEIGHBOR_HOODS_4 = True
OFFSETS_4 = [[0, -1], [-1, 0], [0, 0], [1, 0], [0, 1]]

NEIGHBOR_HOODS_8 = False
OFFSETS_8 = [[-1, -1], [0, -1], [1, -1],
             [-1,  0], [0,  0], [1,  0],
             [-1,  1], [0,  1], [1,  1]]


# 找到每个连通域的索引
def reorganize(binary_img):
    index_map = []
    points = []
    index = -1
    rows, cols = binary_img.shape
    for row in range(rows):
        for col in range(cols):
            val = binary_img[row, col]
            if val == 0:
                continue
            if val in index_map:
                index = index_map.index(val)
                num = index + 1
            else:
                index = len(index_map)
                num = index + 1
                index_map.append(val)
                points.append([])
            binary_img[row, col] = num
            points[index].append([row, col])
    return binary_img, points


def neighbor_value(binarg_img, offsets, reverse=False):
    rows, cols = binary_img.shape
    label_idx = 0
    rows_ = [0, rows, 1] if not reverse else [rows-1, -1, -1]
    cols_ = [0, cols, 1] if not reverse else [cols-1, -1, -1]
    for row in range(rows_[0], rows_[1], rows_[2]):
        for col in range(cols_[0], cols_[1], cols_[2]):
            label = 256
            if binary_img[row, col] == 0:
                continue

            for offset in offsets:
                neighbor_row = min(max(0, row+offset[0]), rows-1)
                neighbor_col = min(max(0, col+offset[1]), cols-1)
                neighbor_val = binarg_img[neighbor_row, neighbor_col]
                if neighbor_val == 0:
                    continue
                label = neighbor_val if neighbor_val < label else label

            if label == 255:
                label_idx += 1
                label = label_idx
            binarg_img[row, col] = label
    return binarg_img


def two_pass(binary_img, neighbor_hoods):
    if neighbor_hoods == NEIGHBOR_HOODS_4:
        offsets = OFFSETS_4
    else:
        offsets = OFFSETS_8
    binary_img = neighbor_value(binary_img, offsets, False)
    binary_img = neighbor_value(binary_img, offsets, True)
    return binary_img


def recursive_seed(binary_img: np.array, seed_row, seed_col, offsets, num, max_num=100):
    rows, cols = binary_img.shape
    binary_img[seed_row][seed_col] = num
    for offset in offsets:
        neighbor_row = min(max(0, seed_row + offset[0]), rows - 1)
        neighbor_col = min(max(0, seed_col + offset[1]), cols - 1)
        val = binary_img[neighbor_row, neighbor_col]
        if val < max_num:
            continue
        binary_img = recursive_seed(binary_img, neighbor_row, neighbor_col, offsets, num, max_num)
    return binary_img


# max_num 表示连通域最多存在的个数, num-1为连通域个数
def seed_filling(binary_img, neighbor_hoods, max_num=100):
    if neighbor_hoods == NEIGHBOR_HOODS_4:
        offsets = OFFSETS_4
    else:
        offsets = OFFSETS_8

    num = 1
    rows, cols = binary_img.shape
    for row in range(rows):
        for col in range(cols):
            val = binary_img[row][col]
            if val <= max_num:
                continue
            binary_img = recursive_seed(binary_img, row, col, offsets, num, max_num=100)
            num += 1
    return binary_img


if __name__ == "__main__":
    binary_img = np.zeros((4, 7), dtype=np.int16)
    index = [[0, 2], [0, 5],
            [1, 0], [1, 1], [1, 2], [1, 4], [1, 5], [1, 6],
            [2, 2], [2, 5],
            [3, 1], [3, 2], [3, 4], [3, 6]]
    for i in index:
        binary_img[i[0], i[1]] = np.int16(255)

    print("原始二值图像")
    print(binary_img)

    # print("Two_Pass")
    # binary_img = two_pass(binary_img, NEIGHBOR_HOODS_8)
    # binary_img, points = reorganize(binary_img)
    # print(binary_img, points)

    print("Seed_Filling")
    binary_img = seed_filling(binary_img, NEIGHBOR_HOODS_8)
    binary_img, points = reorganize(binary_img)
    print(binary_img, points)

File: DengJH/cpp/ConnectedComponent.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

bool NEIGHBOR_HOODS_4 = true;
vector<vector<int> > OFFSETS_4(5, vector<int>(2, 0));
void show_img(vector<vector<int> >& binary_img);

vector<vector<int> >& neighbor_value(vector<vector<int> >& binary_img, vector<vector<int> > offsets, bool reverse)
{
    int rows = binary_img.size();
    int cols = binary_img[0].size();

    int label_index = 0;
    int rows_[2];
    int cols_[2];
    if(!reverse)
    {
        rows_[0] = 0;   rows_[1] = 1;
        cols_[0] = 0;   cols_[1] = 1;
    }
    else
    {
        rows_[0] = rows-1;  rows_[1] = -1;
        cols_[0] = cols-1;  cols_[1] = -1;
    }
    for(int i=rows_[0]; (i<=rows-1) && i>=0; i+=rows_[1])
    {
        for(int j=cols_[0]; (j<=cols-1)&&j>=0; j+=cols_[1])
        {
            if(binary_img[i][j]==0)
                continue;
            int label = 256;   
            for(int k=0; k<offsets.size(); k++)
            {
                int r = min(max(0, i + offsets[k][0]), rows-1);
                int c = min(max(0, j + offsets[k][1]), cols-1);
                int v = binary_img[r][c];
                if(v == 0)
                    continue;
                label = (v < label) ? v : label;
            }
            // 如果周围没有更小的像素就赋新的标签
            if(label == 255)
            {
                label_index++;
                label = label_index;
            }
            binary_img[i][j] = label;
        }
    }
    return binary_img;
}

vector<vector<int> >& two_pass(vector<vector<int> >& binary_img, const vector<vector<int> > offsets)
{
    binary_img = neighbor_value(binary_img, offsets, false);
    show_img(binary_img);

    binary_img = neighbor_value(binary_img, offsets, true);
    return binary_img;
}

vector<vector<int> >& recursive_seed(vector<vector<int> >& binary_img, int seed_row, int seed_col, int num, int max_num=100)
{
    int rows = binary_img.size();
    int cols = binary_img[0].size();
    binary_img[seed_row][seed_col] = num;
    for(int i=0; i<OFFSETS_4.size(); i++)
    {
        int r = min(max(0, seed_row + OFFSETS_4[i][0]), rows-1);
        int c = min(max(0, seed_col + OFFSETS_4[i][1]), cols-1);
        int v = binary_img[r][c];
        if(v < max_num)
            continue;
        binary_img = recursive_seed(binary_img, r, c, num);
    }
    return binary_img;
}

vector<vector<int> >& seed_filling(vector<vector<int> >& binary_img,int max_num=100)
{
    int rows = binary_img.size();
    int cols = binary_img[0].size();
    int num = 1;
    for(int i=0; i<rows; i++)
    {
        for(int j=0; j<cols; j++)
        {
            int val = binary_img[i][j];
            if(val <= max_num)
                continue;
            
            // 只对255的像素值进行递归
            binary_img = recursive_seed(binary_img, i, j, num);
            num++;
        }
    }
    return binary_img;
}

void show_img(vector<vector<int> >& binary_img)
{
    int rows = binary_img.size();
    int cols = binary_img[0].size();
    for(int i=0; i<rows; i++)
    {
        for(int j=0; j<cols; j++)
        {
            cout << binary_img[i][j] << "    ";
        }
        cout << endl;
    }
}

int main()
{
    OFFSETS_4[0] = vector<int>({0, -1});
    OFFSETS_4[1] = vector<int>({-1, 0});
    OFFSETS_4[2] = vector<int>({0, 0});
    OFFSETS_4[3] = vector<int>({0, 1});
    OFFSETS_4[4] = vector<int>({1, 0});

    vector<vector<int> > binary_img(4, vector<int>(7, 0));
    int n = 14;
    int idxs[] = {0, 2, 0, 5,
            1, 0, 1, 1, 1, 2, 1, 4, 1, 5, 1, 6,
            2, 2, 2, 5,
            3, 1, 3, 2, 3, 4, 3, 6};
    for(int i=0; i<n; i++)
    {
        int idx[] = {idxs[2*i], idxs[2*i+1]};
        binary_img[idx[0]][idx[1]] = 255;
    }

    show_img(binary_img);
    // binary_img = two_pass(binary_img, OFFSETS_4);
    binary_img = seed_filling(binary_img);
    show_img(binary_img);
    return 0;
}